packages/gatsby/src/redux/persist.ts

```diff
@@ -1,12 +1,155 @@
+  }
+
+  // Max size of a Buffer is 2gb (yeah, we're assuming 64bit system)
+  // https://stackoverflow.com/questions/8974375/whats-the-maximum-size-of-a-node-js-buffer
+  // Use 1.5gb as the target ceiling, allowing for some margin of error
+  return Math.floor((150 * 1024 * 1024 * 1024) / maxSize)
+}
+
+function prepareCacheFolder(
+  targetDir: string,
+  contents: ICachedReduxState
+): void {
+  // Temporarily save the nodes and remove them from the main redux store
+  // This prevents an OOM when the page nodes collectively contain to much data
+  const map = contents.nodes
+  contents.nodes = undefined
+  writeFileSync(reduxRestFile(targetDir), v8.serialize(contents))
+  // Now restore them on the redux store
+  contents.nodes = map
+
+  if (map) {
+    // Now store the nodes separately, chunk size determined by a heuristic
+    const values: [string, IReduxNode][] = [...map.entries()]
+    const chunkSize = guessSafeChunkSize(values)
+    const chunks = Math.ceil(values.length / chunkSize)
+
+    for (let i = 0; i < chunks; ++i) {
+      writeFileSync(
+        reduxChunkFilePrefix(targetDir) + i,
+        v8.serialize(values.slice(i * chunkSize, i * chunkSize + chunkSize))
+      )
+    }
+  }
+}
+
+function safelyRenameToBak(reduxCacheFolder: string): string {
+  // Basically try to work around the potential of previous renamed caches
+  // not being removed for whatever reason. _That_ should not be a blocker.
+  const tmpSuffix = `.bak`
+  let suffixCounter = 0
+  let bakName = reduxCacheFolder + tmpSuffix // Start without number
+  while (existsSync(bakName)) {
+    ++suffixCounter
+    bakName = reduxCacheFolder + tmpSuffix + suffixCounter
+  }
+  renameSync(reduxCacheFolder, bakName)
+
+  return bakName
+}
+
+export function writeToCache(contents: ICachedReduxState): void {
+  // Note: this should be a transactional operation. So work in a tmp dir and
+  // make sure the cache cannot be left in a corruptable state due to errors.
+
+  const tmpDir = mkdtempSync(`reduxcache`) // linux / windows
+
+  prepareCacheFolder(tmpDir, contents)
+
+  // Replace old cache folder with new. If the first rename fails, the cache
+  // is just stale. If the second rename fails, the cache is empty. In either
+  // case the cache is not left in a corrupt state.
+
+  let bakName = ``
+  if (existsSync(reduxCacheFolder)) {
+    // Don't drop until after swapping over (renaming is less likely to fail)
+    bakName = safelyRenameToBak(reduxCacheFolder)
+  }
 
-export const readFromCache = (): ICachedReduxState =>
-  v8.deserialize(readFileSync(file()))
+  // The redux cache folder should now not exist so we can rename our tmp to it
+  renameSync(tmpDir, reduxCacheFolder)
 
-export const writeToCache = (contents: ICachedReduxState): void => {
-  writeFileSync(file(), v8.serialize(contents))
+  // Now try to yolorimraf the old cache folder
+  try {
+    if (existsSync(legacyCacheFile)) {
+      removeSync(legacyCacheFile)
+    }
```

Comment:
Worth trying to avoid `existSync` here if we already successfully removed this before? Or it pretty much doesn't matter given rest that is happening here and not worth extra (minor) complexity?

Reply:
Technically the first check it doesn't check whether this file exists but rather whether the redux folder exists and then tries to find the legacy cache file.

> Or it pretty much doesn't matter given rest that is happening here and not worth extra (minor) complexity?

Yeah